### PR TITLE
Constructed Payloads

### PR DESCRIPTION
**Pull request by jacobkeeler**
Implementation of [SDL-0052](https://github.com/smartdevicelink/sdl_core/issues/1536)

### CODE REVIEW

File: src/components/protocol_handler/src/protocol_handler_impl.cc [hunk 1]
```diff
@@ -235,6 +298,52 @@ void ProtocolHandlerImpl::SendStartSessionNAck(ConnectionID connection_id,
...
+                                           protocol_version,
+                                           PROTECTION_OFF,
+                                           FRAME_TYPE_CONTROL,
+                                           service_type,
+                                           FRAME_DATA_START_SERVICE_NACK,
+                                           session_id,
+                                           0u,
+                                           message_counters_[session_id]++));
+
+  if (rejectedParams.size() > 0) {
```
Comment (JackLivio) [inline src/components/protocol_handler/src/protocol_handler_impl.cc]: Does there need to be a check for protocol version >= 5 like in StartSessionAck()?

File: src/components/protocol_handler/src/protocol_handler_impl.cc [hunk 2]
```diff
@@ -262,6 +371,51 @@ void ProtocolHandlerImpl::SendEndSessionNAck(ConnectionID connection_id,
...
+      new protocol_handler::ProtocolPacket(connection_id,
+                                           protocol_version,
+                                           PROTECTION_OFF,
+                                           FRAME_TYPE_CONTROL,
+                                           service_type,
+                                           FRAME_DATA_END_SERVICE_NACK,
+                                           session_id,
+                                           0u,
+                                           message_counters_[session_id]++));
+  if (rejectedParams.size() > 0) {
```
Comment (JackLivio) [inline src/components/protocol_handler/src/protocol_handler_impl.cc]: Does there need to be a check for protocol version >= 5 like in StartSessionAck()?

File: src/components/protocol_handler/src/protocol_handler_impl.cc [hunk 3]
```diff
@@ -1121,7 +1326,26 @@ RESULT_CODE ProtocolHandlerImpl::HandleControlMessageStartSession(
                           PROTECTION_OFF);
       return RESULT_OK;
     }
-    if (ssl_context->IsInitCompleted()) {
+    ProtocolPacket::ProtocolVersion* fullVersion;
+    std::vector<std::string> rejectedParams(0, std::string(""));
+    if (packet.data_size() != 0) {
```
Comment (JackLivio) [inline src/components/protocol_handler/src/protocol_handler_impl.cc]: Can you add a comment saying a control packet with data must be at least protocol version 5?

File: src/appMain/smartDeviceLink.ini [hunk 4]
```diff
@@ -222,6 +222,22 @@ AppIconsFolderMaxSize = 104857600
 ; Amount of oldest icons to remove in case of max folder size was reached
 AppIconsAmountToRemove = 1
 
+[SDL5]
+; Enables SDL protocol 5.0 support
+EnableProtocol5 = true
```
Comment (JackLivio) [inline src/appMain/smartDeviceLink.ini]: Maybe refactor this later ^

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 5]
```diff
@@ -1082,7 +1082,8 @@ class ApplicationManagerImpl
     bool operator()(const ApplicationSharedPtr app) const {
       return app
                  ? handle_ == app->device() &&
-                       ProtocolVersion::kV4 == app->protocol_version()
+                       ProtocolVersion::kV4 <= app->protocol_version() &&
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/application_manager_impl.h]: I took it as AppV4 or greater. It looks like it is used to determine what apps support V4 features, so this seems like the way to do it. I would create an AppV5 predicate if one was needed anywhere, but it shouldn't be.

File: src/components/application_manager/include/application_manager/message.h [hunk 6]
```diff
@@ -63,7 +63,8 @@ enum ProtocolVersion {
   kV1 = 1,
   kV2 = 2,
   kV3 = 3,
-  kV4 = 4
+  kV4 = 4,
+  kV5 = 5
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/message.h]: Issue was already entered [here](https://github.com/smartdevicelink/sdl_core/issues/1698), though I will see if I can include in this PR easily

File: src/components/protocol_handler/src/protocol_handler_impl.cc (hunk 3 above)
Comment (jacobkeeler) [inline src/components/protocol_handler/src/protocol_handler_impl.cc]: Done

File: src/components/connection_handler/src/connection.cc [hunk 7]
```diff
@@ -352,8 +352,7 @@ bool Connection::SupportHeartBeat(uint8_t session_id) {
   }
   Session& session = session_it->second;
   return (
-      (::protocol_handler::PROTOCOL_VERSION_3 == session.protocol_version ||
-       ::protocol_handler::PROTOCOL_VERSION_4 == session.protocol_version) &&
+      (session.protocol_version >= ::protocol_handler::PROTOCOL_VERSION_3) &&
```
Comment (jacobkeeler) [inline src/components/connection_handler/src/connection.cc]: Done

File: src/3rd_party/CMakeLists.txt [hunk 8]
```diff
@@ -283,9 +283,27 @@ if(HMI_DBUS_API)
...
+set(BSON_LIBS_DIRECTORY ${3RD_PARTY_INSTALL_PREFIX_ARCH}/lib PARENT_SCOPE)
+set(BSON_INCLUDE_DIRECTORY ${3RD_PARTY_INSTALL_PREFIX}/include PARENT_SCOPE)
+set(EMHASHMAP_INCLUDE_DIRECTORY ${3RD_PARTY_INSTALL_PREFIX}/include PARENT_SCOPE)
+set(EMHASHMAP_LIBS_DIRECTORY ${3RD_PARTY_INSTALL_PREFIX_ARCH}/lib PARENT_SCOPE)
+set(BSON_LIB_SOURCE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/bson_c_lib)
+
+include(ExternalProject)
+ExternalProject_Add(libbson
+    GIT_REPOSITORY "http://github.com/smartdevicelink/bson_c_lib.git"
+    GIT_TAG "master"
```
Comment (LuxoftAKutsan) [inline src/3rd_party/CMakeLists.txt]: @jacobkeeler are you sure that we need always use the latest version of this package? There may be some issues due changes in bson_c_lib master branch

**Comment by LuxoftAKutsan** [comment]
After merging this pull request I have error during incremental build:
```
[ 26%] Performing update step for 'libbson'
Current branch master is up to date.
No stash found.
No stash found.
No stash found.
CMake Error at /home/akutsan/dev/sdl/build/src/3rd_party/libbson-prefix/tmp/libbson-gitupdate.cmake:136 (message):
  

  Failed to unstash changes in:
  '/home/akutsan/dev/sdl/build/src/3rd_party/bson_c_lib/'.

  You will have to resolve the conflicts manually


src/3rd_party/CMakeFiles/libbson.dir/build.make:94: recipe for target 'src/3rd_party/libbson-prefix/src/libbson-stamp/libbson-update' failed
make[2]: *** [src/3rd_party/libbson-prefix/src/libbson-stamp/libbson-update] Error 1
CMakeFiles/Makefile2:328: recipe for target 'src/3rd_party/CMakeFiles/libbson.dir/all' failed
make[1]: *** [src/3rd_party/CMakeFiles/libbson.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
2   0.76s user 0.21s system 35% cpu 2.727 total make
VLC media player 2.2.2 Weatherwax (revision 2.2.2-0-g6259d80)

```

**Pull request merged**